File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/AppMediaPlayer.java
package com.github.anrimian.musicplayer.data.controllers.music.players;

import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.player.error.ErrorType;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;

import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public interface AppMediaPlayer {

    Observable<PlayerEvent> getEventsObservable();

    void prepareToPlay(CompositionSource composition,
                       long startPosition,
                       @Nullable ErrorType previousErrorType);

    void stop();

    void resume();

    void pause();

    void seekTo(long position);

    void setVolume(float volume);

    Observable<Long> getTrackPositionObservable();

    Single<Long> getTrackPosition();

    Single<Long> seekBy(long millis);

    void setPlaySpeed(float speed);

    float getPlaySpeed();

    void release();

    Observable<Boolean> getSpeedChangeAvailableObservable();

}

File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/AppMediaPlayer.kt
package com.github.anrimian.musicplayer.data.controllers.music.players

import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance
import com.github.anrimian.musicplayer.domain.models.player.events.MediaPlayerEvent
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single

interface AppMediaPlayer {

    fun prepareToPlay(
        source: CompositionContentSource,
        previousException: Exception? = null
    ): Completable

    fun stop()

    fun resume()

    fun pause()

    fun release()

    fun seekTo(position: Long)

    fun setVolume(volume: Float)

    fun setSoundBalance(soundBalance: SoundBalance)

    fun getTrackPosition(): Single<Long>

    fun getDuration(): Single<Long>

    fun setPlaybackSpeed(speed: Float)

    fun getTrackPositionObservable(): Observable<Long>

    fun getPlayerEventsObservable(): Observable<MediaPlayerEvent>

    fun getSpeedChangeAvailableObservable(): Observable<Boolean>

}